components/Forms/CreateResourceForm/CreateResourceForm.js

```diff
@@ -0,0 +1,172 @@
+import React from 'react';
+import { func, arrayOf, shape, string, number } from 'prop-types';
+import { Formik, Field } from 'formik';
+import * as Yup from 'yup';
+import { insertIf } from '@innocuous/functions';
+import isEmpty from 'lodash/isEmpty';
+import get from 'lodash/get';
+import { createResource } from 'common/constants/api';
+import { getServerErrorMessage } from 'common/utils/api-utils';
+import { validationErrorMessages } from 'common/constants/messages';
+import Button from 'components/Buttons/Button/Button';
+import Form from 'components/Form/Form';
+import Input from 'components/Form/Input/Input';
+import Select from 'components/Form/Select/Select';
+import styles from './CreateResourceForm.module.css';
+
+// This exists only because we don't have a radio component and I wanna get this done quickly!
+// TODO: Make radio input component
+const free = { value: 'true', label: 'Free' };
+const paid = { value: 'false', label: 'Paid' };
+
+const costOptions = [free, paid];
+
+const validationSchema = Yup.object().shape({
+  name: Yup.string().required(validationErrorMessages.required).default(''),
+  url: Yup.string().required(validationErrorMessages.required).default(''),
+  category: Yup.string().required(validationErrorMessages.required).default(''),
+  languages: Yup.array().default([]),
+  cost: Yup.string().default(free.value),
+  notes: Yup.string().default(''),
+});
+
+CreateResourceForm.propTypes = {
+  categoryOptions: arrayOf(shape({ name: string, id: number })).isRequired,
+  languageOptions: arrayOf(shape({ name: string, id: number })).isRequired,
+  onSuccess: func.isRequired,
+  onFailure: func.isRequired,
+};
+
+function CreateResourceForm({
+  categoryOptions: _categoryOptions,
+  languageOptions: _languageOptions,
+  onSuccess,
+  onFailure,
+}) {
+  const initialValues = React.useMemo(() => validationSchema.cast(), []);
+
+  const categoryOptions = _categoryOptions.map(({ name }) => ({
+    value: name.toLowerCase(),
+    label: name,
+  }));
+
+  const languageOptions = _languageOptions.map(({ name }) => ({
+    value: name.toLowerCase(),
+    label: name,
+  }));
+
+  const handleSubmit = async ({ cost, languages, ...rest }, actions) => {
+    const values = {
+      ...rest,
+      free: cost === free.value,
+      ...insertIf(!isEmpty(languages), { languages }),
+    };
+
+    try {
+      await createResource(values);
+
+      actions.setSubmitting(false);
+      actions.resetForm();
+
+      onSuccess();
+    } catch (error) {
+      actions.setSubmitting(false);
+
+      const resourcesAPIErrors = get(error, 'response.data.errors');
+
+      if (Array.isArray(resourcesAPIErrors)) {
+        const [firstError] = resourcesAPIErrors;
+
+        if (firstError['invalid-params'] && firstError['invalid-params'].params.includes('url')) {
+          onFailure('A resource with this URL already exists in our database!');
+        }
```

Comment:
It may be worth some energy thinking about the best way to handle this.

We don't do field-level server-side error messages, so I'll just be picking the first error I see in the response and trying to parse it.

Otherwise we'll give a generic: "Try again later 🤷‍♂️"

Reply:
The error messages returned from the API are all created with the user in mind, I don't think there's anything wrong with just passing that error through to the user rather than hardcoding one on the front end.

To answer your question though, I think that there will only be one error if there's only one resource submitted. Until we support submitting multiple resources with one request, we can kick that can down the road I would think. And we may never support uploading multiple resources in one request on the front end, that would be ok too.

Reply:
The only problem I have with simply passing through the error is that it's not really language an end-user should be seeing:

```The following params were invalid: url. Resource id 2995 already has this URL.```

I'd rather see:

```A resource with this URL already exists in our database.```

Reply:
Ok, fair enough. We need to keep that ID in there for programmers to see when things go wrong as they're developing on top of the API, so I wouldn't make that change on the backend.

Feel free to leave it like it is 😄